File: HomeAutomation/HomeAutomation/Components/MotionDetector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using HomeAutomation.Core;

namespace HomeAutomation.Components
{
    public class MotionDetector : IComponent
    {
        public bool MotionDetected { get; private set; }
        public DateTime LastMotionTime { get; private set; }
        public string LastLocation { get; private set; }

        public void OnStateChanged(EnvironmentalState state)
        {
            if (state.MotionDetected)
            {
                MotionDetected = true;
                LastMotionTime = DateTime.Now;
                LastLocation = "Hallway";
            }
            else
            {
                MotionDetected = false;
            }
        }
    }
}

File: HomeAutomation/HomeAutomation/src/main/java/com/homeautomation/components/MotionDetector.java
package com.homeautomation.components;

import com.homeautomation.core.Component;
import com.homeautomation.core.EnvironmentalState;

public class MotionDetector implements Component {
    private boolean motionDetected;
    private String lastDetectedLocation;

    public MotionDetector() {
        this.motionDetected = false;
        this.lastDetectedLocation = "None";
    }

    @Override
    public void updateState(EnvironmentalState state) {
        this.motionDetected = state.isMotionDetected();
        this.lastDetectedLocation = state.getMotionLocation();
    }

    public boolean isMotionDetected() {
        return motionDetected;
    }

    public String getLastDetectedLocation() {
        return lastDetectedLocation;
    }
}
